### ISSUE TITLE
Conditional types don't narrow base types

### ISSUE THREAD

**Issue by falsandtru** [Microsoft/TypeScript#23019]
So probably no way to make an index type of mapped types using narrowed types.

cc @ahejlsberg 

**TypeScript Version:**  master

**Search Terms:** 

**Code**

```ts
type A<T> = [string] extends [T] ? Record<T, any> : never;
type B<T> = [string] extends [T] ? [T] extends [string] ? { [i: T]: any; } : never : never;
```

**Expected behavior:**

pass

**Actual behavior:**

```
$ node built/local/tsc.js --noEmit index.ts
index.ts(1,43): error TS2344: Type 'T' does not satisfy the constraint 'string'.
index.ts(2,62): error TS1337: An index signature parameter type cannot be a union type. Consider using a mapped object type instead.
```

**Playground Link:**  

**Related Issues:** #23016

**Comment by falsandtru** [comment]
@ahejlsberg Can you provide a way to make an index type of mapped types using type narrowing by conditional types?

**Comment by falsandtru** [comment]
Note: In the second case, we can directly write the index type as `string`.

**Comment by ahejlsberg** [comment]
No, there is no way to do that. The type specified in an index signature must be either `string` or `number`. I'd consider using a mapped type instead.

**Comment by falsandtru** [comment]
Thanks for the answer. Another, will you provide type narrowing for base types (right-hand of `extends`)? This issue is opened for that.
